trl/trainer/ppov2_trainer.py

```diff
@@ -0,0 +1,819 @@
+        )
+        policy.generation_config.pad_token_id = None  # generate tokens without truncation / padding
+
+        #########
+        # setup model, optimizer, and others
+        #########
+        if args.truncate_token and args.truncate_token == "eos":
+            args.truncate_token_id = tokenizer.eos_token_id
+        self.model = PolicyAndValueWrapper(policy, value_model)
+        self.create_optimizer_and_scheduler(num_training_steps=args.num_updates)
+
+        #########
+        ### trainer specifics
+        #########
+        self.state = TrainerState(
+            is_local_process_zero=self.is_local_process_zero(),
+            is_world_process_zero=self.is_world_process_zero(),
+        )
+        DEFAULT_CALLBACKS = [DefaultFlowCallback]
+        default_callbacks = DEFAULT_CALLBACKS + get_reporting_integration_callbacks(self.args.report_to)
+        if self.callbacks is None:
+            self.callbacks = default_callbacks
+        self.callback_handler = CallbackHandler(
+            self.callbacks, self.model, self.tokenizer, self.optimizer, self.lr_scheduler
+        )
+        self.control = TrainerControl()
+        self.is_deepspeed_enabled = getattr(self.accelerator.state, "deepspeed_plugin", None) is not None
+        self.is_fsdp_enabled = getattr(self.accelerator.state, "fsdp_plugin", None) is not None
+        # Create distant repo and output directory if needed
+        self.hub_model_id = None
+        if self.args.push_to_hub:
+            self.init_hf_repo()
+        if self.args.should_save:
+            os.makedirs(self.args.output_dir, exist_ok=True)
+        self.backup_model = None
+
+        #########
+        ### setup dataloader
+        #########
+        self.dataloader = DataLoader(
+            self.train_dataset,
+            batch_size=args.local_batch_size,
+            shuffle=True,
+            collate_fn=DataCollatorWithPadding(tokenizer),
+            drop_last=True,  # needed; otherwise the last batch will be of ragged shape
+        )
+        # sync random states for DataLoader(shuffle=True) before `accelerator.prepare`
+        # see https://gist.github.com/vwxyzjn/2581bff1e48e185e0b85b6dfe1def79c
+        torch.manual_seed(args.seed)
+        self.model, self.optimizer, self.dataloader = accelerator.prepare(self.model, self.optimizer, self.dataloader)
+        torch.manual_seed(self.local_seed)  # reset the local seed again
+
+        self.eval_dataloader = DataLoader(
+            self.eval_dataset,
+            batch_size=args.per_device_eval_batch_size,
+            collate_fn=DataCollatorWithPadding(self.tokenizer),
+            drop_last=True,
+        ) # no need to shuffle eval dataset
+        self.eval_dataloader = accelerator.prepare(self.eval_dataloader)
+
+    def get_train_dataloader(self) -> DataLoader:
+        return self.dataloader
+    
+    def get_eval_dataloader(self) -> DataLoader:
+        return self.eval_dataloader
+
+    def push_to_hub(self, **kwargs):
```

Comment:
I am not sure I follow. What does having a simple backup mean?

Reply:
Sorry I wasn't clear. What I mean is that we should borrow the code below

```
        backup_model = self.model
        self.model = self.accelerator.unwrap_model(self.model).policy # save only the policy
        super().push_to_hub(**kwargs)
        self.model = backup_model
```

And do something similar for `save_model`, e.g.

```
    def save_model(self, **kwargs):
        backup_model = self.model
        self.model = self.accelerator.unwrap_model(self.model).policy # save only the policy
        super().save_model(**kwargs)
        self.model = backup_model
```

I might be misunderstanding the behavior but it seems because `save_model` calls `push_to_hub` and not the other way around, this is fine.